components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/DefaultLoginValidationDelegate.kt

```diff
@@ -24,27 +24,73 @@ class DefaultLoginValidationDelegate(
     private val scope: CoroutineScope = CoroutineScope(IO)
 ) : LoginValidationDelegate {
 
-    @Suppress("ComplexMethod") // This method is not actually complex
-    override fun validateCanPersist(login: Login): Deferred<Result> {
+    /**
+     * Queries underlying storage for [Login] to determine if it should be updated or created.
+     */
+    @Suppress("ComplexMethod")
+    override fun getNewLoginPersistenceAsync(newLogin: Login): Deferred<Result> {
         return scope.async {
-            try {
-                // We're setting guid=null here in order to trigger de-duping logic properly.
-                // Internally, the library ensures to not dupe records against themselves
-                // (via a `guid <> :guid` check), which means we need to "pretend" this is a new record,
-                // in order for `ensureValid` to actually throw `DUPLICATE_LOGIN`.
-                storage.value.ensureValid(login.copy(guid = null))
-                Result.CanBeCreated
+            val potentialDupes = storage.value.getPotentialDupesIgnoringUsername(newLogin)
+            var shouldCreate = false
+            val loginToValidate = when {
+                // If list is empty, no records match, we know we're creating
+                potentialDupes.isEmpty() -> {
+                    shouldCreate = true
+                    newLogin
+                }
+                // If no usernames match in list, this is a new username, we know we're creating
+                !potentialDupes.any { it.username.isBlank() || it.username == newLogin.username } -> {
+                    shouldCreate = true
+                    newLogin
+                }
+                // Otherwise there's a matching record, let's merge with new password to update
+                else -> {
+                    potentialDupes.first().mergeWithLogin(newLogin)
```

Comment:
This isn't quite right.

At this point, you know that:
- there _could_ be a record with an exact username match; if it exists, you should pick that record for updating (there could be more than one of these, too, in theory - but you can't depend on that being true or not, since then you'd be assuming behaviour details/restrictions internal to the underlying storage layer)
- otherwise, there _could_ be a record with a blank username and the same `guid` as `newLogin`. if it exists, you should pick that record for updating. Ideally, we'd have a separate `Result` for this case - desktop displays a "Add username to this login?" prompt

If none of those are true, you know that list of dupes isn't empty, there aren't any exact username matches, and there are no guid-matching records with a blank username.
If there are any non-guid-matching blank-username records, you could pick one of them for updating, to reduce number of total records present. `guid` could be mismatching for all sorts of sync-related reasons.

Otherwise, it's a "create a new record" case.